### ISSUE TITLE
Force user to accept a privacy policy when creating a profile

### ISSUE THREAD

**Issue by kewisch** [indico/indico#5923]
**Is your feature request related to a problem? Please describe.**
We'd like to set expectations when the user signs up, and give them the chance to bail before a profile is created if they aren't happy with the terms. We need the user to confirm a privacy policy beforehand.

**Describe the solution you'd like**
On https://indico.instance/register/ssoprovider I'd like there to be a checkbox that asks the user to confirm an configurable text with a checkbox, which in my case I'd use to link a privacy policy. I will likely need the flexibility to set arbitrary text and links, as it is important to write down what they are confirming.

The field should be sent to the server and verified there, so that people can't simply say they've disabled the checkbox and have not accepted the terms.

**Describe alternatives you've considered**
I'd like to simply do this in a plugin, since I imagine there might be a number of additional things that different instance owners might want to have on the signup page, but I haven't found a way to actually extend React forms and components from a plugin (something I really hate about React and webpack, previously coming from the browser extensions world. Hope I am just wrong!) 

**Additional context**

![image](https://github.com/indico/indico/assets/607198/0d8959b1-0a3a-4969-929c-a1cb3789cd53)

**Comment by ThiefMaster** [comment]
Adding it from a plugin would be very tricky (the backend of the signup page is also a bit messy), but I think a setting to require agreeing to some condition (just a generic link or text that can be displayed I'd say) when creating an account would be a nice feature for the core. So we certainly wouldn't be opposed to a PR! :)

**Comment by ThiefMaster** [comment]
FWIW if you do want to add it via a plugin, I think creating a react entry point (similar to signals, search for `renderPluginComponents` in the code to find it) to add custom stuff to the signup page would be fine.

The main problem is that currently we do not have anything that would let you nicely inject new fields into an existing marshmallow schema (`SignupSchema` in this case). But what you can do, is using the `schema_post_load` signal and raise a validation error from there if the option is not checked ([example](https://github.com/indico/indico-plugins-cern/blob/master/foundationsync/indico_foundationsync/plugin.py#L67-L93)). However, maybe this isn't even necessary: A client-side required field is more than enough, you do not really need server-side validation whether the user ticked the checkbox, if you have a required `FinalCheckbox` it will already prevent the user from submitting the form unless it's checked.

**Comment by kewisch** [comment]
Sounds good, let's go with client side first. That also makes the patch much more simple. I guess users could force creation without accepting, but then they've actively circumvented measures :)

How is this in /admin/legal ?

![image](https://github.com/indico/indico/assets/607198/5044e38c-53dd-4f68-980e-a41da52d78e1)

**Comment by ThiefMaster** [comment]
I think it's confusing to have it in the same section. Probably better to make it a new section and allow link/text like for the general one

**Comment by kewisch** [comment]
Since the text in that new field will go into the checkbox label raw (with sanitization), I'm not sure that would work. Here is what it looks like with that config:

![image](https://github.com/indico/indico/assets/607198/16c59b49-09c4-4e06-a67c-1330dce856b2)

If there is link/text, then in the link case I'd need to add a generic "I accept the <a href="/tos">privacy policy</a>" label, while with text it would simply display the text as it does now.  I think just having a text field with a default value makes more sense there?

**Comment by kewisch** [comment]
Thought we might not need a default value, depending on if you'd like the checkbox shown by default or not. I can finish things up and send a PR, this might make it easier to play around with and get a feel for it.

**Comment by ThiefMaster** [comment]
I would use the `<Markdown>` component for displaying it

**Comment by kewisch** [comment]
Sounds good, I went with a `<Markdown>` and a simple field without the ckeditor. See #5925 for more details.

**Comment by kewisch** [comment]
fwiw I'm considering a separate PR on top of this that would slightly change the behavior so that a user account has a date field at which they've accepted the TOS, and there is an admin date field where the TOS is last updated (which can be manually bumped).

When the user's date is lower than the admin field, on the next request they are redirected to a page where they are prompted to accept the TOS. This in addition to having the checkbox as is in the new profile.

This way we can cover both the new account case, and site owners wanting to do a major update to the terms that users need to accept, or if they implement the TOS after the fact and want to get buy-in from existing rs.

**Comment by ThiefMaster** [comment]
I think this is a very good idea, feel free to just update this PR.

Using the date may lead to slightly cumbersome UI though. How about a "ToS version" instead, and an `accepted_tos_version` in the User model? If the user's accepted version is lower than the current tos version, they have to approve it again (whether that's required for existing users or not should be another (bool) setting though in case you want to update the ToS but not immediately ask users for new approval).

**Comment by kewisch** [comment]
Heh, ok. I was reluctant to add this to the PR since it is a little more UI work for the admin settings, page to prompt for TOS, etc. I'll see when I have some free time to look into this.

The date works as a mechanism where you can update but not ask immediately, e.g. you could set a date when it goes into effect that is in the future. We used a date setting for this kind of thing for addons.mozilla.org, with the advantage that you could email folks the terms will change, set the date they go into effect, and then not worry about it any more. It would also allow showing a nice "last updated" timestamp in the terms. Numeric value will always be immediate.

I can play around with using a numeric value though when I implement and see how it feels. I don't think it would be much different if it is a number setting or a date setting. A date might be easier to understand even, since you can set the date it should go into effect, as opposed to thinking about version numbers that always need to increment.

**Comment by kewisch** [comment]
Playing around with this I still think a date field works better. Incrementing the numeric value feels very technical. I was also thinking about instead having a checkbox a la "this is a major policy update, reprompt users", which is better than exposing the number, but that still means the policy goes into effect immediately. If I have to do this based on numbers I'd go with that approach though.

I'm not quite as comfortable with the changes so far because I feel there are many unanswered questions that I don't have an immediate answer to, so some screenshots inbetween.

What to display on the actual agreement page?
* We have the TOS, the PP, the text that I intended people to see when they register
* If the TOS and PP are text it seems like a no-brainer to show them inline on the agreements page, in scrollable boxes, as below.

* If PP or TOS are URLs, then we should link to them somehow.
  * We can use hardcoded text instead of custom text and link to TOS and PP depending on what is set. This removes flexibility to a degree, but is mitigated by the fact that admins can still use a text agreement that includes a further link.
  * We can trust the admin to be smart and ensure the custom text include a link to the TOS/PP if they only set URLs. This has the risk that someone might set URLs, but not include them in the prompt text, with the result that the prompt would be meaningless towards the PP/TOS.
  * We can add some validation in the admin that requires there to be a link to /tos and /privacy in the custom text if the URLs are set, though this can be finicky and less flexible. 
  * ... ?

Here is my current agreement page that shows up when the user needs to accept terms again after they registered. I'd add some introductory text a la "The terms have changed, to continue using the platform please accept":

<img width="1115" alt="image" src="https://github.com/indico/indico/assets/607198/10523f96-dc9a-46d9-9f9d-841a48a675ff">

Here are the current admin settings. Depending on how the custom text is used I'd reword that first textbox so it applies to both agreements page and registration page.

<img width="820" alt="image" src="https://github.com/indico/indico/assets/607198/2e3799ea-1091-469f-89aa-6e019514a0b7">

**Comment by ThiefMaster** [comment]
The main reason I don't like the date (and in particular not the automated updates of the date when making changes) is that you might fix a typo or formatting - things that (should) have no legal meaning and you certainly do not want to force an update. With a version number you would simply not bump it.

But maybe we should consider something more fancy altogether. The current implementation was meant to be pretty simple because that's all we needed so far.

Let's say we have a table that stores ToS, Privacy Policy, etc. Compared to the key-value-store (aka SettingsProxy) we currently use, that would allow us to have revisions. And for each user you could reference the revision they accepted. When making a change to either document, you would be able to choose whether this is a "trivial" change (not sure what's the best wording for it) or not. In case of a trivial change, it would simply update the text of the existing version. In case of a non-trivial change it would create a new table row, making it a new revision - which would require users to agree to the updated version.

Instead of going by whether the user agreed to the current revision there could also be flag indicating whether a revision is still active. As long as a user has accepted a revision that is active, they're fine. Once the revision is marked as inactive, they'd need to accept the latest active one. This would for example mean that you can adjust some terms for new users, but not force old users to switch to the newer terms. I cannot imagine why you would need this, just saying that this would be easily possible with such a system.

**Comment by kewisch** [comment]
I was thinking about the revisions case as well, but was also concerned I would be exploding scope for this more. If we do revisions, I'd need to include a full revisions management UI, which is getting quickly out of hand for what I had imagined. I think we should leave this to the next willing soul.

The date wouldn't automatically update, if you want to do typo fixes you simply keep the date as it is:

1) Trivial change: user updates the TOS/PP, leaves the registration text as is, does not bump the date
2) Major change: user updates TOS/PP, can probably leave registration text as is, bumps the date to when it goes into effect.

I'm more concerned about making clear to the admin how all of these things interconnect. We have the TOS, PP, registration text, and some form of versioning. The TOS and PP are already used in a number of places (e.g. footer), the prompt text is yet another privacy thing to keep in mind, and would also be used on two different pages. Registration page is also not easy to preview.

**Comment by ThiefMaster** [comment]
One this I would certainly get rid of is the time component. Just date is more than enough for this.

**Comment by kewisch** [comment]
Ok, sleeping over this I have a plan:

* Use a hardcoded text for the "I accept" checkbox along the lines of "I have read, understood, and accept the terms set forth by the terms of service and privacy policy"
* If there is a text PP or TOS, show it inline, both on the profile creation and agreement page. If someone like me wants a custom text but otherwise linking to another page, I can add that as a PP/TOS text and add a link there.
* If the PP or TOS is an url, link them in the hardcoded text


This should cover all the use cases while keeping it simple. This means no extra text in the settings for the prompt. To make it easier to understand I'd likely add a checkbox "Require users to accept the terms", which would simply disable/enable the registration prompt and agreement page as necessary.

On the admin side, if we get rid of the time component, what timezone would we use for acceptance? If we use local time, then folks could delay the term changes by changing timezones. If we use UTC, it could be a earlier than the admin might expect. For addons.mozilla.org we did have a time component, and usually set this to 12:00pm UTC the next day, so that even folks in UTC+12 would not get the term changes until the date, but we had the administrative/legal certainty that the term changes are atomic, e.g. there is only one time when they start applying to everyone.

**Comment by ThiefMaster** [comment]
Default timezone of the Indico instance maybe?

**Comment by kewisch** [comment]
That sounds right, going with that. Here is a screenshot of the upcoming registration page (plus a last updated line with the date, but I was too lazy to redo the screenshot):

![image](https://github.com/indico/indico/assets/607198/c6b03ced-5c98-48d3-b041-1e47ae0fea30)


Screenshot with TOS text and a PP url. Similarly, if both TOS and PP are an URL, it will just show the checkbox and both items linked.

![image](https://github.com/indico/indico/assets/607198/d50626bd-6776-41e2-b0c5-cc3d36b6731a)

Here is the agreement page that pops up if the user needs to confirm a second time:

![image](https://github.com/indico/indico/assets/607198/72c61ed9-3800-4eaa-9067-6759eb62cc8d)

The agreement works with a redirect that is checked on each request, of course I've excluded certain requests like admin pages or assets. What I've noticed though, if the agreement needs to be accepted in the moment the user is doing something, they might get some unexpected behavior.

For example, if they click on "Create event", then instead of the actual create event dialog showing up, it would redirect and hence put the agreement text into the dialog.

I'm not sure we have a graceful way to handle that across the platform? I've noticed this as well during our work on the multipass groups, if the login expires at the wrong time, some xhr endpoints think something else is wrong and just show an error instead of reloading the rest of the page to trigger a login and then going back to where it left off.

**Comment by ThiefMaster** [comment]
I haven't checked most of the code yet but I suggest to use the same logic as in `_handle_insecure_password_logins` does to determine whether to redirect to the re-acceptance page or not.

Alternatively only checking it during login would be an option, but depending on the session duration maybe not a great one...

**Comment by kewisch** [comment]
Yeah, the session duration is by default pretty much infinite for frequent users. IIRC it was set to 30 days, and it extends itself if the user uses Indico. I didn't have to log in again until we recently did some changes to accommodate for groups expiry. I'd rather do this on each request.

I'm using the same approach as _handle_insecure_password_logins in that I am using `@app.before_request`, and this works very well for the usual case, but since it is a redirect, it will also redirect xhr calls that are made as part of the page loading dialogs. From a user experience perspective we could maybe always allow requests that are xhr, but to an extent this means someone could be using parts of the platform without having accepted updated terms. 

Fixing this might be useful in a different issue though, since it doesn't only affect this redirect. It will also affect expired logins where an xhr is done on page, in which case some pages give you generic errors like "the thing doesn't exist anymore".

Thanks for the review(s)!
